zkevm-circuits/src/keccak_circuit/util.rs

```diff
@@ -462,3 +483,188 @@ pub(crate) fn extract_field<F: FieldExt>(value: Value<F>) -> F {
+        for (idx, expected) in [(0, 0), (1, 1), (128, msb), (129, msb | 1)] {
+            let packed: F = pack(&into_bits(&[idx as u8]));
+            assert_eq!(packed, F::from(expected));
+        }
+    }
+
+    #[test]
+    fn num_bits_per_lookup() {
+        // Typical values.
+        assert_eq!(get_num_bits_per_lookup_impl(3, 19), 11);
+        assert_eq!(get_num_bits_per_lookup_impl(4, 19), 9);
+        assert_eq!(get_num_bits_per_lookup_impl(5, 19), 8);
+        assert_eq!(get_num_bits_per_lookup_impl(6, 19), 7);
+        // The largest imaginable value does not overflow u64.
+        assert_eq!(get_num_bits_per_lookup_impl(3, 32) * BIT_COUNT, 60);
+    }
+
+    #[test]
+    fn normalize_table() {
+        normalize_table_impl(3, 10);
+        normalize_table_impl(4, 10);
+        normalize_table_impl(6, 10);
+        normalize_table_impl(6, 19);
+    }
+
+    fn normalize_table_impl(range: usize, log_height: usize) {
+        let table = build_table(&TableTestCircuit {
+            range,
+            log_height,
+            normalize_else_chi: true,
+        });
+
+        // On all rows, all inputs/outputs are correct, i.e. they have the same low bit.
+        assert_eq!(BIT_COUNT, 3);
+        for (inp, out) in table.iter() {
+            for pos in (0..64).step_by(BIT_COUNT) {
+                assert_eq!((inp >> pos) & 1, (out >> pos) & (4 + 2 + 1));
+            }
+        }
+    }
+
+    #[test]
+    fn chi_table() {
+        // Check the base pattern for all combinations of bits.
+        for i in 0..16_usize {
+            let (a, b, c, d) = (i & 1, (i >> 1) & 1, (i >> 2) & 1, (i >> 3) & 1);
+            assert_eq!(
+                CHI_BASE_LOOKUP_TABLE[3 - 2 * a + b - c],
+                (a ^ ((!b) & c)) as u8
+            );
+            assert_eq!(
+                CHI_EXT_LOOKUP_TABLE[5 - 2 * a - b + c - 2 * d],
+                (a ^ ((!b) & c) ^ d) as u8
+            );
+        }
+
+        // Check the table with multiple parts per row.
+        chi_table_impl(10);
+        chi_table_impl(19);
+    }
+
+    fn chi_table_impl(log_height: usize) {
+        let range = 5; // CHI_BASE_LOOKUP_RANGE
+        let table = build_table(&TableTestCircuit {
+            range,
+            log_height,
+            normalize_else_chi: false,
+        });
+
+        // On all rows, all input/output pairs match the base table.
+        for (inp, out) in table.iter() {
+            for pos in (0..64).step_by(BIT_COUNT) {
+                let inp = ((inp >> pos) & 7) as usize;
+                let out = ((out >> pos) & 7) as u8;
+                assert_eq!(out, CHI_BASE_LOOKUP_TABLE[inp]);
+            }
+        }
+    }
+
+    // ---- Helpers ----
+
+    fn build_table(circuit: &TableTestCircuit) -> Vec<(u64, u64)> {
+        let prover = MockProver::<F>::run(circuit.log_height as u32, circuit, vec![]).unwrap();
+
+        let columns = prover.fixed();
+        assert_eq!(columns.len(), 2);
+        let unused_rows = 6;
```

Comment:
for this magic number, either add some comments, or use ConstrainSystem.mininum_rows() ?

Reply:
done